Realm.Shared/RealmObjectOps.cs

```diff
@@ -0,0 +1,393 @@
+        }
+
+        public static void SetBooleanValue(TableHandle table, RowHandle row, IntPtr columnIndex, bool value)
+        {
+            NativeTable.set_bool(table, columnIndex, row.RowIndex, MarshalHelpers.BoolToIntPtr(value));
+        }
+
+        public static void SetNullableBooleanValue(TableHandle table, RowHandle row, IntPtr columnIndex, bool? value)
+        {
+            if (value.HasValue)
+                NativeTable.set_bool(table, columnIndex, row.RowIndex, MarshalHelpers.BoolToIntPtr(value.Value));
+            else
+                NativeTable.set_null(table, columnIndex, row.RowIndex);
+        }
+
+        public static void SetDateTimeOffsetValue(TableHandle table, RowHandle row, IntPtr columnIndex, DateTimeOffset value)
+        {
+            var marshalledValue = value.ToRealmUnixTimeMilliseconds();
+            NativeTable.set_timestamp_milliseconds(table, columnIndex, row.RowIndex, marshalledValue);
+        }
+
+        public static void SetNullableDateTimeOffsetValue(TableHandle table, RowHandle row, IntPtr columnIndex, DateTimeOffset? value)
+        {
+            if (value.HasValue)
+            {
+                var marshalledValue = value.Value.ToRealmUnixTimeMilliseconds();
+                NativeTable.set_timestamp_milliseconds(table, columnIndex, row.RowIndex, marshalledValue);
+            }
+            else
+                NativeTable.set_null(table, columnIndex, row.RowIndex);
+        }
+
+        public static void SetObjectValue<T>(Realm realm, TableHandle table, RowHandle row, IntPtr columnIndex, T value) where T : RealmObject
+        {
+            if (value == null)
+            {
+                NativeTable.clear_link(table, columnIndex, row.RowIndex);
+            }
+            else
+            {
+                if (!value.IsManaged)
+                    realm.Manage(value);
+                NativeTable.set_link(table, columnIndex, row.RowIndex, value.RowHandle.RowIndex);
+            }
+        }
+
+        public static void SetByteArrayValue(TableHandle table, RowHandle row, IntPtr columnIndex, byte[] value)
+        {
+            if (value == null)
+            {
+                NativeTable.set_null(table, columnIndex, row.RowIndex);
+            }
+            else if (value.Length == 0)
+            {
+                // empty byte arrays are expressed in terms of a BinaryData object with a dummy pointer and zero size
+                // that's how core differentiates between empty and null buffers
+                NativeTable.set_binary(table, columnIndex, row.RowIndex, (IntPtr)0x1, IntPtr.Zero);
+            }
+            else
+            {
+                unsafe
+                {
+                    fixed (byte* buffer = value)
+                    {
+                        NativeTable.set_binary(table, columnIndex, row.RowIndex, (IntPtr)buffer, (IntPtr)value.LongLength);
+                    }
+                }
+            }
+        }
```

Comment:
I don't think adding a lot of APIs on dynamic objects is a good idea because as soon as the C# `dynamic` keyword comes into play you lose all code-completion and it'll be hard to discover those members.

I think there's a sort of a fundamental difference between how dynamic is implemented in the Java binding where you get an object with a dictionary-like API and how dynamic is usually done in C# with the `dynamic` keyword.